query-node/mappings/src/workingGroup.ts

```diff
@@ -0,0 +1,180 @@
+  applicationIdToWorkerIdMap: ApplicationIdToWorkerIdMap
+): Promise<void> {
+  const storageProviderIds = [...applicationIdToWorkerIdMap.values()]
+
+  for (const storageProviderId of storageProviderIds) {
+    // load storage provider
+    const storageProvider = await db.get(StorageProvider, { where: { id: storageProviderId.toString() } as FindConditions<StorageProvider> })
+
+    // reactivate storage provider if it already exists
+    if (storageProvider) {
+      await reactivateStorageProvider(db, storageProvider)
+      continue
+    }
+
+    // create new storage provider
+    const newStorageProvider = new StorageProvider({
+      id: storageProviderId.toString(),
+      type: storageProviderType,
+      isActive: true,
+    })
+
+    await db.save<StorageProvider>(newStorageProvider)
+  }
+
+  // emit log event
+  logger.info("Storage provider has beed created", {ids: storageProviderIds.map(item => item.toString())})
+}
+
+export async function workingGroup_WorkerStorageUpdated(db: DatabaseManager, storageProviderId: WorkerId, newMetadata: Bytes): Promise<void> {
+  // load storage provider
+  const storageProvider = await db.get(StorageProvider, { where: { id: storageProviderId.toString() } as FindConditions<StorageProvider> })
+
+  // ensure storageProvider exists
+  if (!storageProvider) {
+    return inconsistentState('Non-existing storage provider update requested', storageProviderId)
+  }
+
+  storageProvider.metadata = newMetadata.toUtf8()
+
+  await db.save<StorageProvider>(storageProvider)
+
+  // emit log event
+  logger.info("Storage provider has been updated", {id: storageProviderId})
+}
+
+export async function workingGroup_TerminatedWorker(db: DatabaseManager, storageProviderId: WorkerId): Promise<void> {
+  // do removal logic
+  await deactivateStorageProvider(db, storageProviderId)
+
+  // emit log event
+  logger.info("Storage provider has beed removed (worker terminated)", {id: storageProviderId})
+}
+
+export async function workingGroup_WorkerExited(db: DatabaseManager, storageProviderId: WorkerId): Promise<void> {
+  // do removal logic
+  await deactivateStorageProvider(db, storageProviderId)
+
+  // emit log event
+  logger.info("Storage provider has beed removed (worker exited)", {id: storageProviderId})
+}
+
+/////////////////// Helpers ////////////////////////////////////////////////////
+
+async function reactivateStorageProvider(db: DatabaseManager, storageProvider: StorageProvider) {
+  storageProvider.isActive = true
+
+  await db.save<StorageProvider>(storageProvider)
+}
+
+async function deactivateStorageProvider(db: DatabaseManager, storageProviderId: WorkerId) {
+  // load storage provider
+  const storageProvider = await db.get(StorageProvider, { where: { id: storageProviderId.toString() } as FindConditions<StorageProvider> })
+
+  // ensure storageProvider exists
+  if (!storageProvider) {
+    return inconsistentState('Non-existing storage provider deletion requested', storageProviderId)
```

Comment:
message can be improved and generalized to a worker not storage provider

Reply:
Messages has been improved.